igvc_navigation/src/mapper/octomapper.cpp

```diff
@@ -0,0 +1,215 @@
+  }
+
+  // Traverse entire tree
+  std::vector<std::vector<float>> odds_sum(
+      static_cast<unsigned long>(map_options_.lengthGrid()),
+      std::vector<float>(static_cast<unsigned long>(map_options_.widthGrid()), map_options_.default_logodds));
+  for (octomap::OcTree::iterator it = pc_map_pair.octree->begin(), end = pc_map_pair.octree->end(); it != end; ++it)
+  {
+    // If this is a leaf at max depth, then only update that node
+    if (it.getDepth() == pc_map_pair.octree->getTreeDepth())
+    {
+      auto [x, y] = toMapCoordinates(it.getX(), it.getY());
+      if (x < map_options_.lengthGrid() && y < map_options_.widthGrid())
+      {
+        odds_sum[x][y] += it->getLogOdds();
+      }
+      else
+      {
+        ROS_ERROR_STREAM_THROTTLE_NAMED(10, "Point Outside", "Point outside!");
+      }
+    }
+    else
+    {
+      // This isn't a leaf at max depth. Time to iterate
+      int grid_num = 1u << (pc_map_pair.octree->getTreeDepth() - it.getDepth());
+      auto [x, y] = toMapCoordinates(it.getX(), it.getY());
+      for (int dx = 0; dx < grid_num; dx++)
+      {
+        for (int dy = 0; dy < grid_num; dy++)
+        {
+          if (x < map_options_.lengthGrid() && y < map_options_.widthGrid())
+          {
+            odds_sum[x + dx][y + dy] += it->getLogOdds();
+          }
+          else
+          {
+            ROS_ERROR_STREAM_THROTTLE_NAMED(10, "Point Outside", "Point outside!");
+          }
+        }
+      }
+    }
+  }
+
+  // Transfer from log odds to normal probability
+  for (int i = 0; i < map_options_.lengthGrid(); i++)
+  {
+    for (int j = 0; j < map_options_.widthGrid(); j++)
+    {
+      pc_map_pair.map->at<uchar>(i, j) = toCharProb(fromLogOdds(odds_sum[i][j]));
+    }
+  }
+}
+
+void Octomapper::create_map(pc_map_pair &pair) const
+{
+  int length = static_cast<int>(map_options_.lengthGrid());
+  int width = static_cast<int>(map_options_.widthGrid());
+  pair.map = boost::make_shared<cv::Mat>(length, width, map_encoding_, 127);
+}
+
+void Octomapper::insertScan(const tf::Point &sensor_pos, struct pc_map_pair &pair, const PointCloud &pc,
+                            ProbabilityModel model, double range) const
+{
+  double old_prob_hit = pair.octree->getProbHit();
+  double old_prob_miss = pair.octree->getProbMiss();
+
+  pair.octree->setProbHit(model.prob_hit);
+  pair.octree->setProbMiss(model.prob_miss);
+
+  octomap::point3d sensor = octomap::pointTfToOctomap(sensor_pos);
+  sensor.z() = 0;
+  octomap::Pointcloud octo_cloud;
+  PCL_to_Octomap(pc, octo_cloud);
+  pair.octree->insertPointCloud(octo_cloud, sensor, range, false, false);  // No discretize, no lazy eval
+
+  pair.octree->setProbHit(old_prob_hit);
+  pair.octree->setProbMiss(old_prob_miss);
+}
+
+void Octomapper::insertRays(const tf::Point &sensor_pos, struct pc_map_pair &pair, const PointCloud &pc, bool occupied,
+                            ProbabilityModel model) const
+{
+  double old_prob_hit = pair.octree->getProbHit();
```

Comment:
What is the point of changing the hit probability then reverting it back to its previous value?

Reply:
so that different scans can have different probabilities, ie a different one for lidar scans, free space etc. By default, only one hit / miss probability is used.

Reply:
Understood